src/functions.jl

```diff
@@ -278,6 +278,109 @@ function Base.copy(f::ScalarQuadraticFunction)
     )
 end
 
+"""
+    ScalarNonlinearFunction(head::Symbol, args::Vector{Any})
+
+The scalar-valued nonlinear function `head(args...)`, represented as a symbolic
+expression tree, with the call operator `head` and ordered arguments in `args`.
+
+## `head`
+
+The `head::Symbol` must be an operator supported by the model.
+
+The default list of supported univariate operators is given by:
+
+ * [`Nonlinear.DEFAULT_UNIVARIATE_OPERATORS`](@ref)
+
+and the default list of supported multivariate operators is given by:
+
+ * [`Nonlinear.DEFAULT_MULTIVARIATE_OPERATORS`](@ref)
+
+Additional operators can be registered by setting a [`UserDefinedFunction`](@ref)
+attribute.
+
+See the full list of operators supported by a [`ModelLike`](@ref) by querying
+[`ListOfSupportedNonlinearOperators`](@ref).
+
+## `args`
+
+The vector `args` contains the arguments to the nonlinear function. If the
+operator is univariate, it must contain one element. Otherwise, it may contain
+multiple elements. Each element must be one of the following:
+
+ * A constant value of type `T<:Number`
+ * A [`VariableIndex`](@ref)
+ * A [`ScalarAffineFunction`](@ref)
+ * A [`ScalarQuadraticFunction`](@ref)
+ * A [`ScalarNonlinearFunction`](@ref)
+
+## Unsupported operators
+
+If the optimizer does not support `head`, an [`UnsupportedNonlinearOperator`](@ref)
+error will be thrown.
+
+There is no guarantee about when this error will be thrown; it may be thrown
```

Comment:
This is kind of weird from an error handling perspective. Are there other similar cases? Should `optimize!` fail with an `INVALID_MODEL` termination status instead of throwing?

Reply:
This can happen with `UnsupportedConstraint`, which depends whether there is a cache or whether `copy_to` or `add_constraint` is used.

I added `UnsupportedNonlinearOperator` because it's a pretty serious error that can't be recovered from. It's not like an invalid model that needs a starting point changed or at least one constraint.